File: second-year/second-semester/MPP/mpp-proiect-repository-raulbaciulescu/travelAgency/Persistence/src/main/java/impl/database/UserTable.java
package impl.database;

import api.Constants;
import api.Table;
import domain.User;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.*;

public class UserTable implements Table<Long, User> {
    private final Connection connection;
    private final Map<Constants.Db.Queries, PreparedStatement> statements;
    private static final Logger logger = LogManager.getLogger();

    public UserTable(Connection connection) throws SQLException {
        logger.info("Initializing UserTable");
        this.statements = new HashMap<>();
        this.connection = connection;
        initStatements();
    }

    void initStatements() throws SQLException {
        statements.put(Constants.Db.Queries.ADD,
                connection.prepareStatement("INSERT INTO user (id, username, password, firstName, lastName) VALUES (?, ?, ?, ?, ?)"
                ));
        statements.put(Constants.Db.Queries.DELETE,
                connection.prepareStatement("DELETE FROM user WHERE id = ?"
                ));
        statements.put(Constants.Db.Queries.FIND_BY_ID,
                connection.prepareStatement("SELECT * FROM user WHERE id = ?"
                ));
        statements.put(Constants.Db.Queries.GET_ALL,
                connection.prepareStatement("SELECT * FROM user"
                ));
        statements.put(Constants.Db.Queries.FIND2,
                connection.prepareStatement("SELECT * FROM user WHERE username = ? AND password = ?"
                ));

    }
    @Override
    public void add(User user) {
        logger.traceEntry("saving user {} ", user);
        PreparedStatement statement = statements.get(Constants.Db.Queries.ADD);
        try {
            statement.setLong(1, user.getId());
            statement.setString(2, user.getUsername());
            statement.setString(3, user.getPassword());
            statement.setString(4, user.getFirstName());
            statement.setString(5, user.getLastName());
            int result = statement.executeUpdate();
            logger.trace("Saved {} instances", result);
        } catch (SQLException e) {
            logger.error(e);
            System.err.println("Error DB " + e);
        }
        logger.traceExit();
    }

    @Override
    public void delete(Long o) {

    }

    @Override
    public void update(User elem, User newElem) {

    }

    @Override
    public Optional<User> findById(Long id) {
        logger.traceEntry();
        List<User> users = new ArrayList<>();
        PreparedStatement statement = statements.get(Constants.Db.Queries.FIND_BY_ID);
        try {
            statement.setLong(1, id);
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    long id1 = result.getLong("id");
                    String username = result.getString("username");
                    String password = result.getString("password");
                    String firstName = result.getString("firstName");
                    String lastName = result.getString("lastName");
                    User user = new User(username, password, firstName, lastName);
                    user.setId(id1);
                    users.add(user);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        return Optional.of(users.get(0));
    }

    @Override
    public List<User> getAll() {
        logger.traceEntry();
        List<User> users = new ArrayList<>();
        PreparedStatement statement = statements.get(Constants.Db.Queries.GET_ALL);
        try {
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    long id1 = result.getLong("id");
                    String username = result.getString("username");
                    String password = result.getString("password");
                    String firstName = result.getString("firstName");
                    String lastName = result.getString("lastName");
                    User user = new User(username, password, firstName, lastName);
                    user.setId(id1);
                    users.add(user);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        return users;
    }

    public Optional<User> findUser(String username, String password) {
        logger.traceEntry();
        List<User> users = new ArrayList<>();
        PreparedStatement statement = statements.get(Constants.Db.Queries.FIND2);
        try {
            statement.setString(1, username);
            statement.setString(2, password);
            try (ResultSet result = statement.executeQuery()) {
                while (result.next()) {
                    long id1 = result.getLong("id");
                    String username1 = result.getString("username");
                    String password1 = result.getString("password");
                    String firstName = result.getString("firstName");
                    String lastName = result.getString("lastName");
                    User user = new User(username, password, firstName, lastName);
                    user.setId(id1);
                    users.add(user);
                }
            }
        } catch (SQLException throwables) {
            logger.error(throwables);
            System.out.println("Error db" + throwables);
        }
        logger.traceExit();
        return Optional.of(users.get(0));
    }
}

File: second-year/second-semester/MPP/mpp-proiect-repository-raulbaciulescu/travelAgency2/persistance/database/UserTable.cs

using System.Data.SQLite;
using model;
using model.Utils;
using persistance.Database;

namespace travelAgency2.Repository.Database
{
    internal class UserTable : ITable<long, User>
    {
        private SQLiteConnection connection;
        private Dictionary<Db.Queries, SQLiteCommand> statements;
        private static readonly log4net.ILog logger = log4net.LogManager.GetLogger(System.Reflection.MethodBase.GetCurrentMethod().DeclaringType);


        public UserTable(SQLiteConnection connection)
        {
            logger.Info("Initializing UserTable");
            this.connection = connection;
            this.statements = new Dictionary<Db.Queries, SQLiteCommand>();    
            InitStatements();
        }

        void InitStatements()
        {

            SQLiteCommand command1 = new SQLiteCommand(connection);
            SQLiteCommand command2 = new SQLiteCommand(connection);
            SQLiteCommand command3 = new SQLiteCommand(connection);
            SQLiteCommand command4 = new SQLiteCommand(connection);
            SQLiteCommand command5 = new SQLiteCommand(connection);
            command1.CommandText = "INSERT INTO user (id, username, password, firstName, lastName) VALUES (@id, @username, @password, @firstName, @lastName)";
            command2.CommandText = "DELETE FROM user WHERE id = @id";
            command3.CommandText = "SELECT * FROM user WHERE id = @id";
            command4.CommandText = "SELECT * FROM user";
            command5.CommandText = "SELECT * FROM user WHERE username = @username AND password = @password;";
            statements.Add(Db.Queries.ADD, command1);
            statements.Add(Db.Queries.DELETE, command2);
            statements.Add(Db.Queries.FIND_BY_ID, command3);
            statements.Add(Db.Queries.GET_ALL, command4);
            statements.Add(Db.Queries.FIND2, command5);
        }
        public void Update(User user, User userNew)
        {
            //TODO
        }

        public void Add(User user)
        {
            logger.Info("enter in add user");
            SQLiteCommand statement = statements[Db.Queries.ADD];
            try
            {
                statement.Parameters.AddWithValue("@id", user.Id);
                statement.Parameters.AddWithValue("@username", user.username);
                statement.Parameters.AddWithValue("@password", user.password);
                statement.Parameters.AddWithValue("@firstName", user.firstName);
                statement.Parameters.AddWithValue("@lastName", user.lastName);
                int numberofrowsaffected = statement.ExecuteNonQuery();
                String s = String.Format("Saved {0} instances", numberofrowsaffected);
                logger.Info(numberofrowsaffected);
            }
            catch (Exception ex)
            {
                logger.Error(ex);
                Console.WriteLine(ex.Message);
            }
            logger.Info("exit from add user");
        }

        public void Delete(long id)
        {
           //TODO
        }

        public User FindById(long id)
        {
            logger.Info("enter in findById user");
            List<User> users = new List<User>();  
            SQLiteCommand statement = statements[Db.Queries.FIND_BY_ID];
            statement.Parameters.AddWithValue("@id", id);
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    User user = new User((long)reader.GetInt64(4), reader.GetString(0), reader.GetString(1), reader.GetString(2), reader.GetString(3));
                    users.Add(user);
                }
            }
            logger.Info("exit from findById user");
            reader.Close();
            if (users.Count > 0)
                return users[0];
            return null;
        }

        public List<User> GetAll()
        {
            logger.Info("enter in getAll user");
            List<User> users = new List<User>();
            SQLiteCommand statement = statements[Db.Queries.GET_ALL];
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    User user = new User((long) reader.GetInt64(4), reader.GetString(0), reader.GetString(1), reader.GetString(2), reader.GetString(3));
                    users.Add(user);
                }
            }
            reader.Close();
            logger.Info("exit from getAll user");
            return users;
        }
        public User findUser(string username, string password)
        {
            logger.Info("enter in findUser user");
            List<User> users = new List<User>();
            SQLiteCommand statement = statements[Db.Queries.FIND2];
            statement.Parameters.AddWithValue("@username", username);
            statement.Parameters.AddWithValue("@password", password);
            SQLiteDataReader reader = statement.ExecuteReader();
            if (reader.HasRows)
            {
                while (reader.Read())
                {
                    User user = new User((long)reader.GetInt64(4), reader.GetString(0), reader.GetString(1), reader.GetString(2), reader.GetString(3));
                    users.Add(user);
                }
            }
            logger.Info("exit from findUser user");
            reader.Close();
            if (users.Count > 0)
                return users[0];
            return null;
        }
    }
}
